File: src/cpu/probes/MyInstTracker.py
from m5.objects import SimObject
from m5.objects.Probe import ProbeListenerObject
from m5.params import *
from m5.util.pybind import *

class MyGlobalInstTracker(SimObject):
    type = "MyGlobalInstTracker"
    cxx_header = "cpu/probes/my_inst_tracker.hh"
    cxx_class = "gem5::MyGlobalInstTracker"
    cxx_exports = [
        PyBindMethod("changeThreshold"),
        PyBindMethod("resetCounter"),
        PyBindMethod("getThreshold")
    ]
    inst_threshold = Param.Counter("The instruction threshold to trigger an"
                                                                " exit event")
    
class MyLocalInstTracker(ProbeListenerObject):
    type = "MyLocalInstTracker"
    cxx_header = "cpu/probes/my_inst_tracker.hh"
    cxx_class = "gem5::MyLocalInstTracker"
    cxx_exports = [
        PyBindMethod("stopListening"),
        PyBindMethod("startListening")
    ]
    global_inst_tracker = Param.MyGlobalInstTracker("Global instruction tracker")
    start_listening = Param.Counter(True, "Start listening for instructions")

File: src/cpu/probes/my_inst_tracker.cc
#include "cpu/probes/my_inst_tracker.hh"

namespace gem5{

    MyGlobalInstTracker::MyGlobalInstTracker(const MyGlobalInstTrackerParams &params):
        SimObject(params),
        instCount(0),
        instThreshold(params.inst_threshold)
    {}

    void
    MyGlobalInstTracker::checkPc(const uint64_t& inst)
    {
        instCount ++;
        if (instCount >= instThreshold) {
            exitSimLoopNow("a thread reached the max instruction count");
        }
    }

    MyLocalInstTracker::MyLocalInstTracker(const MyLocalInstTrackerParams &params):
        ProbeListenerObject(params),
        globalInstTracker(params.global_inst_tracker),
        listening(params.start_listening)
    {}

    void
    MyLocalInstTracker::regProbeListeners(){
        if (listening) {
            if (listeners.empty()) {
                connectListener<MyLocalInstTrackerListener>(
                    this, "RetiredInsts", &MyLocalInstTracker::checkPc);
                DPRINTF(InstTracker, "Start listening to RetiredInsts\n");
            }
        }
    }

    void
    MyLocalInstTracker::checkPc(const uint64_t& inst)
    {
        globalInstTracker->checkPc(inst);
    }
}